### PR TITLE
Optimize String.escaped

### PR DESCRIPTION
**Pull request by nojb**
A small optimization, avoiding some allocation in the fast path.

The patch is due to @alainfrisch, I just packaged it as a PR.

Do we need a Changes entry?

### CODE REVIEW

File: stdlib/string.ml [hunk 1]
```diff
@@ -100,15 +100,19 @@ let trim s =
...
+let needs_escape s =
+  let n = length s in
+  let i = ref 0 in
+  let esc = ref false in
+  while !i < n && not !esc do
+    match unsafe_get s !i with
+    | '\"' | '\\' | '\000'..'\031' | '\127'.. '\255' -> esc := true
+    | _ -> incr i
+  done;
+  !esc
```
Comment (xavierleroy) [inline stdlib/string.ml]: This "while not !finished do ... done" idiom brings bad memory from Pascal.  What's wrong with the original recursion?  Remember, this is a functional language.

Comment (alainfrisch) [reply stdlib/string.ml]: Speedups come from:

  - A while-loop is still faster than tail-rec loops.
  - Not recomputing length for every character.
  - Patterns more friendly to the PM compiler (I was suprised, but the cmm code for the original code is clearly sub-optimal).

I'll post benchmarks later on.

Comment (xavierleroy) [reply stdlib/string.ml]: Not recomputing the length is fine.  Pascal style is not to be encouraged.

Comment (alainfrisch) [reply stdlib/string.ml]: Ok, so I propose:

````ocaml
let needs_escape s =
  let rec needs_escape s n i =
    if i >= n then false else
      match String.unsafe_get s i with
      | '\"' | '\\' | '\n' | '\t' | '\r' | '\b' -> true
      | ' ' .. '~' -> needs_escape s n (i+1)
      | _ -> true
  in
  needs_escape s (String.length s) 0
`````

which gives most of the gain.

**Comment by gasche** [comment]
When we discussed adapting the standard library to `Bytes/String`, we discussed such cases (also `String.concat` with a single element, etc.), and decided to not perform optimization that would change the "natural" sharing behavior of the function, to not break the (possibly careless) code of -unsafe-string users.

It may be reasonable to open the flood gates now that `-unsafe-string` is not the default anymore, but I think we should discuss this decision globally, not about just a single function. (The discussion can, of course, happen here.)

**Comment by alainfrisch** [comment]
> and decided to not perform optimization that would change the "natural" sharing behavior of the function

This is unrelated to the current PR, which only implements differently the way String.escaped checks for its already-existing fast-path.

**Comment by gasche** [comment]
Oh, I misread, sorry.

File: stdlib/string.ml (hunk 1 above)
Comment (alainfrisch) [reply stdlib/string.ml]: Some very quick testing suggests that we go from about 550 Mb/s for the current version to 715 Mb/s with the version above.

**Review by gasche — APPROVED** [review]
I reviewed the patch and believe that the change is correct. (It's not clear that making it a non-internal function is a readability win, but either ways are fine.)

**Comment by yallop** [comment]
It would be nice if flambda¹ eventually allowed us to avoid hand-optimising loops like this.

For example, perhaps flambda could observe that `length s` and `i` don't change within the loop, and that `needs_escape` doesn't escape, and transform the existing code into the equivalent of the latest version of the code in this PR (b2be2129).

Or, even better, perhaps flambda could allow us to avoid writing special purpose loops with unsafe accesses altogether; we could have a high-level function

```ocaml
val exists : (char -> bool) -> string -> bool
```

and then define `needs_escape` as

```ocaml
exists (function '\"' | '\\' | '\000'..'\031' | '\127'.. '\255' -> true | _ -> false)
```

¹ i.e. the optimising middle-end in general

**Comment by alainfrisch** [comment]
> It would be nice if flambda¹ eventually allowed us to avoid hand-optimising loops like this.

Indeed, but let's not forget Javascript backends...

File: stdlib/string.ml [hunk 2]
```diff
@@ -100,15 +100,17 @@ let trim s =
...
+    if i >= n then false else
+      match String.unsafe_get s i with
+      | '\"' | '\\' | '\000'..'\031' | '\127'.. '\255' -> true
+      | _ -> aux s n (i+1)
   in
-  if needs_escape 0 then
+  aux s (String.length s) 0
+
+let escaped s =
+  if needs_escape s then
```
Comment (alainfrisch) [inline stdlib/string.ml]: On second thought (sorry), the code could be simplified by directly calling from `escaped` the recursive function (with the three arguments).  It probably improves even a bit the generated bytecode.

File: stdlib/string.ml [hunk 3]
```diff
@@ -100,15 +100,17 @@ let trim s =
...
-    if i >= length s then false else
-      match unsafe_get s i with
-      | '\"' | '\\' | '\n' | '\t' | '\r' | '\b' -> true
-      | ' ' .. '~' -> needs_escape (i+1)
-      | _ -> true
+let needs_escape s =
+  let rec aux s n i =
+    if i >= n then false else
+      match String.unsafe_get s i with
+      | '\"' | '\\' | '\000'..'\031' | '\127'.. '\255' -> true
```
Comment (alainfrisch) [inline stdlib/string.ml]: Side note: it is "tempting" to drop the check on the length altogether (also avoiding the need to compute the length), relying on the fact that `String.unsafe_get s (String.length s)` returns `\000`.  The trouble, in addition to stressing @xavierleroy, is that this doesn't work with js_of_ocaml (and presumably Bucklescript).

**Comment by xavierleroy** [comment]
How much time Lexifi's employees can consume to micro-optimizing a non-time-critical library function?

**Comment by alainfrisch** [comment]
We have an actual use case where String.escaped turns out to be called a lot on strings that don't need to be escaped, and where the change discussed here brings an observable speedup (not a factor, but a few percents of the total running time).

**Comment by Octachron** [comment]
Since there is a at least one use case where this change  brings "an observable speedup", does this warrant a Changes entry as an user-visible change?

**Comment by yallop** [comment]
> On second thought (sorry), the code could be simplified by directly calling from escaped the recursive function (with the three arguments).

In a similar vein, you could avoid the branch in `escaped` by returning `bts (B.escaped (bos s))` and `s` directly from `needs_escaped`, rather than returning `true` and `false`.

**Comment by nojb** [comment]
> In a similar vein, you could avoid the branch in escaped by returning bts (B.escaped (bos s)) and s directly from needs_escaped, rather than returning true and false.

That's right, thanks.

**Comment by gasche** [comment]
but then you should really go back to a local function within `escape` :-)

**Comment by nojb** [comment]
> but then you should really go back to a local function within escape :-)

Fair enough, will do :)

File: stdlib/string.ml [hunk 4]
```diff
@@ -101,17 +101,13 @@ let trim s =
   else s
 
 let escaped s =
-  let rec needs_escape i =
-    if i >= length s then false else
+  let rec escape s n i =
```
Comment (gasche) [inline stdlib/string.ml]: Could this function be called `escape_if_needed` for clarity? What it is doing is sort of weird.

(I promise to make this my last nitpicking. On this PR.)

Comment (nojb) [reply stdlib/string.ml]: Done!

**Review by alainfrisch — APPROVED** [review]

**Comment by alainfrisch** [comment]
@nojb Feel free to merge directly after adding the Changes entry.

**Pull request merged**